File: app/src/main/java/com/fengwei23/androidsample/MessageDialogFragment.java
package com.fengwei23.androidsample;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.Fragment;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import com.sunplus.toolbox.utils.BuildCheck;

/**
 * Created by w.feng on 2018/10/11
 * Email: fengweisb@gmail.com
 */
public class MessageDialogFragment extends DialogFragment {
  private static final String TAG = MessageDialogFragment.class.getSimpleName();

  public interface MessageDialogListener {
    void onMessageDialogResult(final MessageDialogFragment dialog, final int requestCode,
                               final String[] permissions, final boolean result);
  }

  public static MessageDialogFragment showDialog(final Activity parent, final int requestCode,
                                                 final int idTitle, final int idMessage,
                                                 final String[] permissions) {
    final MessageDialogFragment dialog =
        newInstance(requestCode, idTitle, idMessage, permissions);
    dialog.show(parent.getFragmentManager(), TAG);
    return dialog;
  }

  public static MessageDialogFragment showDialog(final Fragment parent, final int requestCode,
                                                 final int idTitle, final int idMessage,
                                                 final String[] permissions) {
    final MessageDialogFragment dialog =
        newInstance(requestCode, idTitle, idMessage, permissions);
    dialog.setTargetFragment(parent, parent.getId());
    dialog.show(parent.getFragmentManager(), TAG);
    return dialog;
  }

  public static MessageDialogFragment newInstance(final int requestCode, final int idTitle,
                                                  final int idMessage,
                                                  final String[] permissions) {
    final MessageDialogFragment fragment = new MessageDialogFragment();
    final Bundle args = new Bundle();
    args.putInt("requestCode", requestCode);
    args.putInt("title", idTitle);
    args.putInt("message", idMessage);
    args.putStringArray("permissions", permissions != null ? permissions : new String[] {});
    fragment.setArguments(args);
    return fragment;
  }

  private MessageDialogListener mDialogListener;

  public MessageDialogFragment() {
    super();
  }

  @SuppressLint("NewApi")
  @Override
  public void onAttach(final Activity activity) {
    super.onAttach(activity);
    //获取回调接口
    if (activity instanceof MessageDialogListener) {
      mDialogListener = (MessageDialogListener) activity;
    }
    if (mDialogListener == null) {
      final Fragment fragment = getTargetFragment();
      if (fragment instanceof MessageDialogListener) {
        mDialogListener = (MessageDialogListener) fragment;
      }
    }
    if (mDialogListener == null) {
      if (BuildCheck.isAndroid42()) {
        final Fragment target = getParentFragment();
        if (target instanceof MessageDialogListener) {
          mDialogListener = (MessageDialogListener) target;
        }
      }
    }
    if (mDialogListener == null) {
      throw new ClassCastException(activity.toString());
    }
  }


  @Override
  public Dialog onCreateDialog(final Bundle savedInstanceState) {
    final Bundle args = savedInstanceState != null ? savedInstanceState : getArguments();
    final int requestCode = getArguments().getInt("requestCode");
    final int idTitle = getArguments().getInt("title");
    final int idMessage = getArguments().getInt("message");
    final String[] permissions = args.getStringArray("permissions");

    return new AlertDialog.Builder(getActivity())
        .setIcon(android.R.drawable.ic_dialog_alert)
        .setTitle(idTitle)
        .setMessage(idMessage)
        .setPositiveButton(android.R.string.ok,
            new DialogInterface.OnClickListener() {
              @Override
              public void onClick(final DialogInterface dialog, final int whichButton) {
                try {
                  mDialogListener.onMessageDialogResult(MessageDialogFragment.this, requestCode,
                      permissions, true);
                } catch (final Exception e) {
                  Log.w(TAG, e);
                }
              }
            }
        )
        .setNegativeButton(android.R.string.cancel,
            new DialogInterface.OnClickListener() {
              @Override
              public void onClick(final DialogInterface dialog, int whichButton) {
                try {
                  mDialogListener.onMessageDialogResult(MessageDialogFragment.this, requestCode,
                      permissions, false);
                } catch (final Exception e) {
                  Log.w(TAG, e);
                }
              }
            }
        )
        .create();
  }
}

File: screenrecorderdemo/src/main/java/com/almosting/sample/MessageDialogFragment.kt
package com.almosting.sample

import android.R.drawable
import android.R.string
import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog.Builder
import android.app.Dialog
import android.app.DialogFragment
import android.app.Fragment
import android.os.Bundle
import android.util.Log
import com.almosting.toolbox.utils.BuildCheck

/**
 * Created by w.feng on 2018/10/11
 * Email: fengweisb@gmail.com
 */
class MessageDialogFragment : DialogFragment() {
  interface MessageDialogListener {
    fun onMessageDialogResult(
      dialog: MessageDialogFragment?, requestCode: Int,
      permissions: Array<String?>?, result: Boolean
    )
  }

  private var mDialogListener: MessageDialogListener? = null

  @SuppressLint("NewApi")
  override fun onAttach(activity: Activity) {
    super.onAttach(activity)
    //获取回调接口
    if (activity is MessageDialogListener) {
      mDialogListener = activity
    }
    if (mDialogListener == null) {
      val fragment = targetFragment
      if (fragment is MessageDialogListener) {
        mDialogListener = fragment
      }
    }
    if (mDialogListener == null) {
      if (BuildCheck.isAndroid42()) {
        val target = parentFragment
        if (target is MessageDialogListener) {
          mDialogListener = target
        }
      }
    }
    if (mDialogListener == null) {
      throw ClassCastException(activity.toString())
    }
  }

  override fun onCreateDialog(savedInstanceState: Bundle): Dialog {
    val args =
      savedInstanceState ?: arguments
    val requestCode = arguments.getInt("requestCode")
    val idTitle = arguments.getInt("title")
    val idMessage = arguments.getInt("message")
    val permissions = args.getStringArray("permissions")
    return Builder(activity)
      .setIcon(drawable.ic_dialog_alert)
      .setTitle(idTitle)
      .setMessage(idMessage)
      .setPositiveButton(
        string.ok
      ) { dialog, whichButton ->
        try {
          mDialogListener!!.onMessageDialogResult(
            this@MessageDialogFragment, requestCode,
            permissions, true
          )
        } catch (e: Exception) {
          Log.w(TAG, e)
        }
      }
      .setNegativeButton(
        string.cancel
      ) { dialog, whichButton ->
        try {
          mDialogListener!!.onMessageDialogResult(
            this@MessageDialogFragment, requestCode,
            permissions, false
          )
        } catch (e: Exception) {
          Log.w(TAG, e)
        }
      }
      .create()
  }

  companion object {
    private val TAG = MessageDialogFragment::class.java.simpleName
    fun showDialog(
      parent: Activity, requestCode: Int,
      idTitle: Int, idMessage: Int,
      permissions: Array<String>
    ): MessageDialogFragment {
      val dialog = newInstance(requestCode, idTitle, idMessage, permissions)
      dialog.show(parent.fragmentManager, TAG)
      return dialog
    }

    fun showDialog(
      parent: Fragment, requestCode: Int,
      idTitle: Int, idMessage: Int,
      permissions: Array<String>
    ): MessageDialogFragment {
      val dialog =
        newInstance(requestCode, idTitle, idMessage, permissions)
      dialog.setTargetFragment(parent, parent.id)
      dialog.show(parent.fragmentManager, TAG)
      return dialog
    }

    fun newInstance(
      requestCode: Int, idTitle: Int,
      idMessage: Int,
      permissions: Array<String>
    ): MessageDialogFragment {
      val fragment = MessageDialogFragment()
      val args = Bundle()
      args.putInt("requestCode", requestCode)
      args.putInt("title", idTitle)
      args.putInt("message", idMessage)
      args.putStringArray(
        "permissions",
        permissions
      )
      fragment.arguments = args
      return fragment
    }
  }
}